cmd/infracost/comment_github.go

```diff
@@ -0,0 +1,140 @@
+		Use:   "github",
+		Short: "Post an Infracost comment to GitHub",
+		Long:  "Post an Infracost comment to GitHub",
+		Example: `  Update a comment on a pull request:
+
+      infracost comment github --repo my-org/my-github-repo --pull-request 3 --path infracost.json --github-token $GITHUB_TOKEN
+
+  Post a new comment to a commit:
+
+      infracost comment github --repo my-org/my-github-repo --commit 2ca7182 --path infracost.json --behavior hide-and-new --github-token $GITHUB_TOKEN`,
+		ValidArgs: []string{"--", "-"},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ctx.SetContextValue("platform", "github")
+
+			var err error
+
+			apiURL, _ := cmd.Flags().GetString("github-api-url")
+			token, _ := cmd.Flags().GetString("github-token")
+			tag, _ := cmd.Flags().GetString("tag")
+			extra := comment.GitHubExtra{
+				APIURL: apiURL,
+				Token:  token,
+				Tag:    tag,
+			}
+
+			commit, _ := cmd.Flags().GetString("commit")
+			prNumber, _ := cmd.Flags().GetInt("pull-request")
+			repo, _ := cmd.Flags().GetString("repo")
+
+			var commentHandler *comment.CommentHandler
+			if prNumber != 0 {
+				ctx.SetContextValue("targetType", "pull-request")
+
+				commentHandler, err = comment.NewGitHubPRHandler(ctx.Context(), repo, strconv.Itoa(prNumber), extra)
+				if err != nil {
+					return err
+				}
+			} else if commit != "" {
+				ctx.SetContextValue("targetType", "commit")
+
+				commentHandler, err = comment.NewGitHubCommitHandler(ctx.Context(), repo, commit, extra)
+				if err != nil {
+					return err
+				}
+			} else {
+				ui.PrintUsage(cmd)
+				return fmt.Errorf("either --commit or --pull-request is required")
+			}
+
+			behavior, _ := cmd.Flags().GetString("behavior")
+			if behavior != "" && !contains(validCommentGitHubBehaviors, behavior) {
+				ui.PrintUsage(cmd)
+				return fmt.Errorf("--behavior only supports %s", strings.Join(validCommentGitHubBehaviors, ", "))
+			}
+			ctx.SetContextValue("behavior", behavior)
+
+			paths, _ := cmd.Flags().GetStringArray("path")
+
+			body, err := buildCommentBody(ctx, paths, output.MarkdownOptions{
+				WillUpdate:          prNumber != 0 && behavior == "update",
+				WillReplace:         prNumber != 0 && behavior == "delete-and-new",
+				IncludeFeedbackLink: true,
+			})
+			if err != nil {
+				return err
+			}
+
+			dryRun, _ := cmd.Flags().GetBool("dry-run")
+			if !dryRun {
+				err = commentHandler.CommentWithBehavior(ctx.Context(), behavior, string(body))
+				if err != nil {
+					return err
+				}
+
+				pricingClient := apiclient.NewPricingAPIClient(ctx)
+				err = pricingClient.AddEvent("infracost-comment", ctx.EventEnv())
+				if err != nil {
+					log.Errorf("Error reporting event: %s", err)
+				}
+
+				cmd.Println("Comment posted to GitHub")
+			} else {
+				cmd.Println(string(body))
+				cmd.Println("Comment not posted to GitHub (--dry-run was specified)")
+			}
+
+			if outFile, _ := cmd.Flags().GetString("out-file"); outFile != "" {
```

Comment:
Hmmm.  I guess if you want only the markdown you can use `infracost output --format github-token`... so maybe there's not a strong use case for out-file and we should just remove it for now.  I was thinking it would be nice for testing, but I made the --dry-run print the comment to stdout and that may be enough for tests.

@aliscott any thoughts?

Reply:
The main existing reason for `--dry-run` is for the CI examples E2E tests. 

We want to replace lines like this in the examples:

```sh
script:
  ...
  - infracost comment gitlab --repo my-org/my-gitlab-repo --merge-request 3 --path infracost.json --gitlab-token $GITLAB_TOKEN
```

With something that will output the generated comment so we can compare it to the golden file.

So I think we have a few options:

1. `--dry-run` performs as normal but skips write actions (posting, hiding, deleting comments), and outputs the comment to stdout.
2. Same as 1, but instead of `--dry-run` we use `INFRACOST_COMMENT_DRY_RUN` env var, since there's no user use-case for `--dry-run` yet.
3. Same as 1, but doesn't output the comment, we use a combination of `--dry-run` and `--out-file` for that
4. `--dry-run` Only outputs the comment to stdout

I'm (weakly) leaning towards number 1 or 2. The issue with 4 is that it doesn't test that we pass the correct credentials to the GitHub/GitLab API.

I don't know the use case for `--out-file` on its own yet, so I say we leave it out until we hear from users.

Reply:
good point about no use case for dry-run, I'll make it an env so opt 2